File: pkg/extension/extension.go
// Package extension provides the public Extension interface and ExtensionContext
// that all Beluga extensions implement. This is the stable API for extension authors.
package extension

import (
	"context"
	"encoding/json"
	"log/slog"

	"github.com/aspectrr/beluga/internal/core/database"
	"github.com/aspectrr/beluga/internal/core/eventstore"
	"github.com/aspectrr/beluga/internal/core/model"
	"github.com/aspectrr/beluga/internal/core/session"
	"github.com/aspectrr/beluga/internal/core/tools"
)

// Extension is the interface all extensions implement.
type Extension interface {
	// Name returns the extension's identifier (used in config and logging).
	Name() string

	// Init is called once at startup. Parse config, create clients,
	// register tools in the registry. Extensions that need gRPC
	// register services on ctx.GRPC here.
	Init(ctx ExtensionContext) error

	// Start is called after all extensions have been initialized.
	// Use it to start background goroutines (connectors, listeners).
	// Must block until ctx is cancelled or Stop is called.
	Start(ctx context.Context) error

	// Stop is called on graceful shutdown. Clean up resources.
	Stop(ctx context.Context) error
}

// ExtensionContext is everything an extension gets access to.
type ExtensionContext struct {
	// Config is the raw YAML config for this extension from beluga.yaml.
	Config json.RawMessage

	// Core services the extension can use.
	Registry *tools.Registry   // Register tools here
	Sessions *session.Store    // Create/query sessions
	Events   *eventstore.Store // Append/query events
	DB       *database.ExtDB   // Restricted database access — see database.ExtPermissions
	Docker   interface{}       // *client.Client — use interface{} to avoid docker import
	Logger   *slog.Logger

	// PromptDir is the path to .beluga/prompts/. Extensions can write
	// prompt template files here to inject behavioral context.
	PromptDir string

	// GRPC is nil unless ext_host is enabled. Extensions that need gRPC
	// (like remora) should check this and fail clearly.
	GRPC interface{} // **GRPCProvider — nil unless ext_host extension is enabled

	// CreateSession creates a new agent session from an external event.
	// Connectors call this when they detect a new task/mention/message.
	CreateSession func(ctx context.Context, source, sourceID string, initialMessage string, metadata json.RawMessage) (*model.Session, error)
}

File: packages/sdk/src/extension.ts
// ── Extension interface ────────────────────────────────────────
// Every Beluga extension must implement this interface.

import type { ExtensionContext } from "./context.js";
import type { SandboxRunner } from "./tools.js";

/** Declarative workspace requirements for an extension.
 *  Beluga merges these across all enabled extensions at workspace build time. */
export interface WorkspaceRequirements {
	/** Python packages to pip install */
	python?: string[];
	/** Node.js packages to npm install -g */
	node?: string[];
	/** System (apt) packages to install */
	system?: string[];
	/** Arbitrary RUN commands added to the generated Dockerfile layer */
	run?: string[];
	/** Environment variables to set in the workspace */
	env?: Record<string, string>;
}

export interface Extension {
	/** Unique identifier for this extension */
	name: string;

	/** Declarative workspace requirements. Merged at build time. */
	workspace?: WorkspaceRequirements;

	/** Called once at startup. Register tools, parse config, etc. */
	init(ctx: ExtensionContext): Promise<void>;

	/** Called after all extensions are initialized. Start background work here. */
	start(signal: AbortSignal): Promise<void>;

	/** Called on graceful shutdown. Clean up resources. */
	stop(): Promise<void>;

	/** Called after a workspace container is created for a session.
	 *  Use this to run imperative setup commands inside the container
	 *  (e.g. downloading binaries, generating config, starting services). */
	onWorkspaceReady?(sandbox: SandboxRunner): Promise<void>;
}
